### PR TITLE
transport: enable virtiofs and vhost-user-fs

### PR DESCRIPTION
**Pull request by bergwolf**
Bring back virtiofs and vhost-user-fs support so that consumers can use
the latest code to run virtiofs and vhost-user-fs.

### CODE REVIEW

**Comment by bergwolf** [comment]
Tested with qemu 6.1.0 with arguments like:
```
        -chardev socket,id=foobar,path=./vhost-fs.sock \
        -device vhost-user-fs-pci,chardev=foobar,tag=sharedfs \
```

File: src/transport/virtiofs/mod.rs [hunk 1]
```diff
@@ -99,19 +103,14 @@ pub type Result<T> = std::result::Result<T, Error>;
 
 impl std::error::Error for Error {}
 
-impl<'a, M: GuestMemory> Reader<'a, M> {
+impl<'a> Reader<'a> {
     /// Construct a new Reader wrapper over `desc_chain`.
-    #[allow(clippy::new_ret_no_self)]
-    pub fn new(mem: &'a M, desc_chain: DescriptorChain<'a, M>) -> Result<Reader<'a, MS<'a, M>>> {
+    pub fn new(
+        mem: &'a GuestMemoryMmap,
```
Comment (jiangliu) [inline src/transport/virtiofs/mod.rs:109]: Why hard-coding to "GuestMemoryMmap" and "GuestMemoryAtomic<GuestMemoryMmap>" here?

File: src/transport/virtiofs/mod.rs [hunk 2]
```diff
@@ -437,7 +449,9 @@ mod async_io {
     }
 }
 
-#[cfg(test)]
+/// Disabled since vm-virtio doesn't export any DescriptorChain constructors.
```
Comment (jiangliu) [inline src/transport/virtiofs/mod.rs:452]: I'm trying to export DescriptorChain::new()

File: tests/smoke.rs [hunk 3]
```diff
@@ -3,13 +3,13 @@
...
 
-#[cfg(feature = "fusedev")]
+#[cfg(all(feature = "fusedev", not(feature = "virtiofs")))]
 #[macro_use]
 extern crate log;
 
 mod example;
 
-#[cfg(feature = "fusedev")]
+#[cfg(all(feature = "fusedev", not(feature = "virtiofs")))]
```
Comment (jiangliu) [inline tests/smoke.rs:12]: why disable it for virtiofs?

File: src/transport/virtiofs/mod.rs (hunk 1 above)
Comment (bergwolf) [reply src/transport/virtiofs/mod.rs:109]: I'm getting 
```
the trait `vm_memory::GuestMemory` is not implemented for `GuestMemoryAtomic<GuestMemoryMmap>`
```
although `GuestMemoryMmap` implements `vm_memory::GuestMemory`

There seems to miss an abstraction of `vm_memory::GuestMemory` for `GuestMemoryAtomic` in vm-memory crate.

Besides, this drops the generic type requirement on Reader/Writer and makes caller's life much easier. Is there need to create a generic type based Reader/Writer? The `VhostUserBackend` trait is based on `GuestMemoryAtomic<GuestMemoryMmap>` as well so I'm using the same concrete type.

File: src/transport/virtiofs/mod.rs (hunk 2 above)
Comment (bergwolf) [reply src/transport/virtiofs/mod.rs:452]: Great. We can re-enable these tests once `DescriptorChain::new()` is exported.

File: tests/smoke.rs (hunk 3 above)
Comment (bergwolf) [reply tests/smoke.rs:12]: The example tests are only for `fusedev` and fails to compile when `virtiofs` is enabled.

File: src/transport/virtiofs/mod.rs (hunk 1 above)
Comment (jiangliu) [reply src/transport/virtiofs/mod.rs:109]: It should works with vhost-user-backend, not sure whether it breaks rund or not.

Comment (jiangliu) [reply src/transport/virtiofs/mod.rs:109]: The prototype of Reader::new()/Writer::new() is not well-formed, so cause the compilation failure.

**Review by jiangliu — APPROVED** [review]

File: src/transport/virtiofs/mod.rs (hunk 1 above)
Comment (bergwolf) [reply src/transport/virtiofs/mod.rs:109]: I tried to change it to 
```
 104 impl<'a, M: GuestMemory> Reader<'a, M> {
 105     /// Construct a new Reader wrapper over `desc_chain`.
 106     pub fn new(
 107         mem: &'a M,
 108         desc_chain: DescriptorChain<&M>,
 109     ) -> Result<Reader<'a, MS<'a, M>>> {
```
and got the above errors with adapting nydusd to use it. I think vm-memory should provide a wrapper for GuestMemoryAtomic over GuestMemory trait.

Comment (jiangliu) [reply src/transport/virtiofs/mod.rs:109]: We should use vm-memory in another way, will try to solve it later.

**Pull request merged**